File: clients/go/smoke.go
// Cross-language interop client for the smoke test, built against the ergonomic
// github.com/moq-dev/moq-go wrapper (package moq: Dial, CreateBroadcast,
// PublishMediaStream, SubscribeMedia, range-over-func frame iterators). The raw
// uniffi-bindgen-go surface lives in github.com/moq-dev/moq-go-ffi; this client
// exercises the idiomatic wrapper a real Go user would reach for.
//
// publish reads raw Annex-B H.264 from stdin (e.g. piped from ffmpeg) and feeds
// it to a streaming importer, which infers frame boundaries. subscribe connects,
// finds the video track in the catalog, and exits 0 as soon as any non-empty
// frame arrives (exit 1 on timeout / no data). Usage:
//
//	ffmpeg ... -f h264 - | go-smoke publish --url http://127.0.0.1:4443 --broadcast b.hang
//	go-smoke subscribe --url http://127.0.0.1:4443 --broadcast b.hang --timeout 20
package main

import (
	"context"
	"flag"
	"fmt"
	"io"
	"os"
	"time"

	"github.com/moq-dev/moq-go/moq"
)

const readChunk = 64 * 1024

func main() {
	if len(os.Args) < 2 {
		fmt.Fprintln(os.Stderr, "usage: go-smoke publish|subscribe --url U --broadcast B [--timeout S]")
		os.Exit(2)
	}
	role := os.Args[1]

	fs := flag.NewFlagSet(role, flag.ExitOnError)
	url := fs.String("url", "", "MoQ server URL")
	broadcast := fs.String("broadcast", "", "broadcast name")
	timeout := fs.Float64("timeout", 20, "subscribe timeout in seconds")
	_ = fs.Parse(os.Args[2:])

	if *url == "" || *broadcast == "" {
		fmt.Fprintln(os.Stderr, "error: --url and --broadcast are required")
		os.Exit(2)
	}

	var err error
	switch role {
	case "publish":
		err = publish(*url, *broadcast)
	case "subscribe":
		err = subscribe(*url, *broadcast, *timeout)
	default:
		fmt.Fprintf(os.Stderr, "unknown role: %s\n", role)
		os.Exit(2)
	}

	if err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}
}

func publish(url, broadcast string) error {
	// The relay uses a self-signed cert, so verification is off. With no origin
	// options, Dial shares one origin for publish + consume.
	client, err := moq.Dial(context.Background(), url, moq.WithTLSVerify(false))
	if err != nil {
		return err
	}
	defer client.Close()

	producer, err := client.CreateBroadcast(broadcast)
	if err != nil {
		return err
	}
	defer producer.Finish()

	// avc3: a self-describing Annex-B H.264 stream the importer can frame on its
	// own. PublishMediaStream feeds the raw byte stream; whole frames are emitted
	// as they complete.
	media, err := producer.PublishMediaStream("avc3")
	if err != nil {
		return err
	}
	defer media.Finish()

	fmt.Fprintf(os.Stderr, "publishing %q (Annex-B H.264 from stdin) to %s\n", broadcast, url)

	buf := make([]byte, readChunk)
	for {
		n, rerr := os.Stdin.Read(buf)
		if n > 0 {
			if werr := media.Write(buf[:n]); werr != nil {
				return werr
			}
		}
		if rerr == io.EOF {
			break
		}
		if rerr != nil {
			return rerr
		}
	}
	return media.Finish()
}

func subscribe(url, broadcast string, timeoutS float64) error {
	// The whole subscribe must complete within the timeout; a cancelled context
	// aborts any in-flight wrapper call and unblocks the frame iterator.
	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeoutS*float64(time.Second)))
	defer cancel()

	client, err := moq.Dial(ctx, url, moq.WithTLSVerify(false))
	if err != nil {
		return err
	}
	defer client.Close()

	announced, err := client.AnnouncedBroadcast(broadcast)
	if err != nil {
		return err
	}
	defer announced.Cancel()

	bc, err := announced.Available(ctx)
	if err != nil {
		return err
	}

	name, video, err := videoTrack(ctx, bc)
	if err != nil {
		return err
	}

	media, err := bc.SubscribeMedia(name, video.Container, nil)
	if err != nil {
		return err
	}
	defer media.Cancel()

	for frame, err := range media.Frames(ctx) {
		if err != nil {
			return err
		}
		if len(frame.Payload) > 0 {
			fmt.Fprintf(os.Stderr, "received %d bytes from %q\n", len(frame.Payload), broadcast)
			return nil
		}
	}
	return fmt.Errorf("no frame data received")
}

// videoTrack waits for a catalog update that actually carries a video track. A
// lazy publisher (e.g. the browser, which only encodes on demand) may announce
// video in a later update, not the first snapshot.
func videoTrack(ctx context.Context, bc *moq.BroadcastConsumer) (string, moq.Video, error) {
	cat, err := bc.SubscribeCatalog()
	if err != nil {
		return "", moq.Video{}, err
	}
	defer cat.Cancel()

	for catalog, err := range cat.Updates(ctx) {
		if err != nil {
			return "", moq.Video{}, err
		}
		for name, video := range catalog.Video {
			return name, video, nil
		}
	}
	return "", moq.Video{}, fmt.Errorf("catalog stream ended without a video track")
}

File: clients/python/smoke.py
"""Cross-language interop client for the smoke test (PyPI `moq-rs`, import `moq`).

publish:   read raw Annex-B H.264 from stdin (e.g. piped from ffmpeg) and feed
           it to a streaming importer, which infers frame boundaries.
subscribe: connect, find the video track in the catalog, and exit 0 as soon as
           any non-empty frame arrives (exit 1 on timeout / no data).

    ffmpeg ... -f h264 - | python smoke.py publish --url http://localhost:4443 --broadcast b.hang
    python smoke.py subscribe --url http://localhost:4443 --broadcast b.hang --timeout 20
"""

import argparse
import asyncio
import sys

import moq

READ_CHUNK = 64 * 1024


async def publish(url: str, broadcast: str) -> None:
    async with moq.Client(url, tls_verify=False) as client:
        # create_broadcast registers the broadcast and hands back its producer
        # (the old client.publish(broadcast, producer) split was removed in the
        # ergonomic moq-rs API). publish_media_stream feeds a raw Annex-B pipe to
        # the streaming importer, which infers frame boundaries.
        producer = client.create_broadcast(broadcast)
        media = producer.publish_media_stream("avc3")
        print(f"publishing {broadcast!r} (Annex-B H.264 from stdin) to {url}")

        loop = asyncio.get_running_loop()
        stdin = sys.stdin.buffer
        # read1 returns as soon as any bytes are available (read() would block
        # for a full chunk and batch up ffmpeg's real-time output). getattr both
        # keeps pyright happy (BinaryIO doesn't declare read1) and falls back if
        # the stream lacks it.
        read = getattr(stdin, "read1", stdin.read)
        while True:
            # Blocking read off the event loop so the client keeps flushing.
            chunk = await loop.run_in_executor(None, read, READ_CHUNK)
            if not chunk:
                break
            media.write(chunk)
        media.finish()


async def _catalog_with_video(consumer: moq.BroadcastConsumer) -> moq.Catalog:
    # The catalog is a live track. A lazy publisher (e.g. the browser, which only
    # encodes on demand) may announce video in a *later* update, not the first
    # snapshot, so wait for a catalog that actually has a video track.
    catalog_consumer = await consumer.subscribe_catalog()
    async for catalog in catalog_consumer:
        if catalog.video:
            return catalog
    raise RuntimeError("catalog stream ended without a video track")


async def subscribe(url: str, broadcast: str, timeout: float) -> None:
    async with moq.Client(url, tls_verify=False) as client:
        # announced_broadcast yields a handle; available() resolves it to a live
        # consumer once the broadcast is announced. subscribe_catalog/_media are
        # coroutines in the ergonomic API, so they're awaited (below and above).
        announced = client.announced_broadcast(broadcast)
        consumer = await asyncio.wait_for(announced.available(), timeout)
        catalog = await asyncio.wait_for(_catalog_with_video(consumer), timeout)

        track_name = next(iter(catalog.video))
        video = catalog.video[track_name]

        media = await consumer.subscribe_media(track_name, video.container)

        total = 0

        async def drain() -> None:
            nonlocal total
            async for frame in media:
                total += len(frame.payload)
                if total > 0:
                    return

        await asyncio.wait_for(drain(), timeout)

    if total <= 0:
        raise RuntimeError("no frame data received")
    print(f"received {total} bytes from {broadcast!r}")


def main() -> None:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("role", choices=["publish", "subscribe"])
    parser.add_argument("--url", required=True)
    parser.add_argument("--broadcast", required=True)
    parser.add_argument("--timeout", type=float, default=20.0)
    args = parser.parse_args()

    try:
        if args.role == "publish":
            asyncio.run(publish(args.url, args.broadcast))
        else:
            asyncio.run(subscribe(args.url, args.broadcast, args.timeout))
    except KeyboardInterrupt:
        pass
    except (TimeoutError, asyncio.TimeoutError):
        print("error: timed out waiting for data", file=sys.stderr)
        sys.exit(1)
    except Exception as err:  # noqa: BLE001 - smoke client: any failure is a failure
        print(f"error: {err}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
